src/poly1305.rs

```diff
@@ -16,31 +16,115 @@
+    buf: [u8; BLOCK_STATE_SIZE + 7],
+    inner: &'a mut [u8; BLOCK_STATE_SIZE],
+}
+impl<'a> AlignedBlockState<'a> {
+    pub fn new(inner: &mut [u8; BLOCK_STATE_SIZE]) -> AlignedBlockState {
+        let mut st = AlignedBlockState {
+            buf: [0u8; BLOCK_STATE_SIZE + 7],
+            inner: inner,
+        };
+        AlignedBlockState::align(&mut st.buf).copy_from_slice(&st.inner[..]);
+        st
+    }
+    fn align(xs: &mut [u8; BLOCK_STATE_SIZE + 7]) -> &mut [u8; BLOCK_STATE_SIZE] {
+        let aligned_start = (xs.as_ptr() as usize + 7) & !7;
+        let offset = aligned_start - (xs.as_ptr() as usize);
+        slice_as_array_ref_mut!(&mut xs[offset..offset+BLOCK_STATE_SIZE], BLOCK_STATE_SIZE).unwrap()
+    }
+    pub fn aligned_buf(&mut self) -> &mut [u8; BLOCK_STATE_SIZE] {
+        AlignedBlockState::align(&mut self.buf)
+    }
+}
+impl<'a> Drop for AlignedBlockState<'a> {
+    fn drop(&mut self) {
+        self.inner.copy_from_slice(AlignedBlockState::align(&mut self.buf));
+    }
 }
 
 impl SigningContext {
     #[inline]
     pub fn from_key(key: Key) -> SigningContext {
-        let mut ctx = SigningContext {
-            state: [0u8; STATE_LEN],
+        let mut ctx = SigningContext{
+            opaque: [0; BLOCK_STATE_SIZE],
+            nonce: [0; 4],
+            buf: [0; BLOCK_LEN],
+            buf_used: 0,
+            func: Funcs {
+                blocks_fn: GFp_poly1305_blocks,
+                emit_fn: GFp_poly1305_emit
+            }
         };
-        unsafe { GFp_poly1305_init(&mut ctx.state, &key.bytes); }
+        check_block_state_layout();
+
+        let set_fns = init(AlignedBlockState::new(&mut ctx.opaque).aligned_buf(), &key.bytes, &mut ctx.func) == 0;
+        // TODO XXX: It seems at least some implementations |poly1305_init| always
+        // return the same value, so this conditional logic isn't always necessary.
+        // And, for platforms that have such conditional logic also in the ASM code,
+        // it seems it would be better to move the conditional logic out of the asm
+        // and into the higher-level code.
+        if !set_fns {
+            ctx.func.blocks_fn = GFp_poly1305_blocks;
+            ctx.func.emit_fn = GFp_poly1305_emit;
+        }
+
+        ctx.nonce = [
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[16..20], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[20..24], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[24..28], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[28..32], 4).unwrap()),
+        ];
         ctx
     }
 
-    #[inline]
-    pub fn update(&mut self, input: &[u8]) {
-        unsafe {
-            GFp_poly1305_update(&mut self.state, input.as_ptr(), input.len());
+    pub fn update(&mut self, mut input: &[u8]) {
+        let mut block_state = AlignedBlockState::new(&mut self.opaque);
+        let block_state_buf = block_state.aligned_buf();
```

Comment:
I suggest that we do something like this:
```rust
let mut state = AlignedBlockState::new();
state.load_execute_save(&mut self.opaque, |state| {
    // use the state.
}
```

The first line will fix the address of `state` and the second line would then copy `self.opaque` into `state` at the appropriate offset to align it. Then it would execute the closure, passing it the aligned pointer. Then after the closure completes, it would copy the updated state back into `self.opaque`. This would be equivalent to the `Drop` stuff but IMO much more clear since it is explicit.